### PR TITLE
8247403: JShell: No custom input (e.g. from GUI) possible with JavaShellToolBuilder

### PR DESCRIPTION
**Pull request by lahodaj**
Basically, JShell's ConsoleIOContext always uses the JLine's default terminal, which uses System.in, or equivalent, for input. But, if the JShell tool has been created using JavaShellToolBuilder, a different input might have been provided. The proposed solution is to use a different JLine's terminal in this case, based on the provided input.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8247403](https://bugs.openjdk.java.net/browse/JDK-8247403): JShell: No custom input (e.g. from GUI) possible with JavaShellToolBuilder

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to d7eacd6a8bbd535a352f73fffc579041af0b630f

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2509/head:pull/2509` \
`$ git checkout pull/2509`

Update a local copy of the PR: \
`$ git checkout pull/2509` \
`$ git pull https://git.openjdk.java.net/jdk pull/2509/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2509`

View PR using the GUI difftool: \
`$ git pr show -t 2509`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2509.diff">https://git.openjdk.java.net/jdk/pull/2509.diff</a>

</details>

### CODE REVIEW

**Review by vicente-romero-oracle — APPROVED** [review]
LGTM

**Comment by lahodaj** [comment]
It turned out some users/clients of JavaShellToolBuilder do not attach a full terminal to it, but rather parse the output, and since we have changed the output by using a newer JLine, this became more complicated. So the new iteration of the patch supports two modes: a (default) simpler mode, which does not use escape sequences much, and a more for real terminals, where JShell will use escape sequences to provide a good experiment to a user reading on the terminal. Please also see the associated CSR:
https://bugs.openjdk.java.net/browse/JDK-8265966

**Comment by vicente-romero-oracle** [comment]
in the CSR consider adding in the @implSpec section a mention to that the provided default implementation does something in the lines of:
```
The default implementation of this method returns {@code this}.
```

**Review by vicente-romero-oracle — APPROVED** [review]

File: src/jdk.internal.le/share/classes/jdk/internal/org/jline/reader/impl/LineReaderImpl.java [hunk 1]
```diff
@@ -4157,7 +4157,7 @@ private AttributedString expandPromptPattern(String pattern, int padToWidth,
             } else
                 sb.append(ch);
         }
-        if (padToWidth > cols) {
```
Comment (archiecobbs) [inline src/jdk.internal.le/share/classes/jdk/internal/org/jline/reader/impl/LineReaderImpl.java:4160]: Random blast-from-the-past question: would it be appropriate to submit this change back to the jline project? It doesn't appear to be in the current jline3 master.

**Pull request closed without merging**